src/i2c_bit_bang.cpp

```diff
@@ -0,0 +1,255 @@
+  const auto output_false_ticks = after_output_low - before_output_low;
+
+  m_scl_low_ticks -= output_false_ticks;
+}
+
+void i2c_bit_bang::driver_transaction(
+  hal::byte p_address,
+  std::span<const hal::byte> p_data_out,
+  std::span<hal::byte> p_data_in,
+  function_ref<hal::timeout_function> p_timeout)
+{
+
+  hal::i2c_operation operation;
+  hal::byte address_to_write;
+
+  // select read or write
+  if (!p_data_out.empty()) {
+    send_start_condition();
+    operation = hal::i2c_operation::write;
+    address_to_write = hal::to_8_bit_address(p_address, operation);
+
+    write_address(address_to_write, p_timeout);
+
+    write(p_data_out, p_timeout);
+  }
+
+  // if the first bit is 1 then we read
+  if (!p_data_in.empty()) {
+    send_start_condition();
+
+    operation = hal::i2c_operation::read;
+    address_to_write = hal::to_8_bit_address(p_address, operation);
+
+    write_address(address_to_write, p_timeout);
+
+    read(p_data_in, p_timeout);
+  }
+
+  send_stop_condition();
+}
+
+// private
+
+void i2c_bit_bang::send_start_condition()
+{
+  using namespace std::chrono_literals;
+  // the start condition requires both the sda and scl lines to be pulled high
+  // before sending, so we do that here.
+  m_sda->level(true);
+  m_scl->level(true);
+  high_speed_delay(m_scl_high_ticks);
+  m_sda->level(false);
+  high_speed_delay(m_scl_high_ticks);
+  m_scl->level(false);
+  high_speed_delay(m_scl_high_ticks);
+}
+
+void i2c_bit_bang::send_stop_condition()
+{
+  m_sda->level(false);
+
+  m_scl->level(true);
+  high_speed_delay(m_scl_high_ticks);
+  m_sda->level(true);
+  high_speed_delay(m_scl_high_ticks);
+}
+
+void i2c_bit_bang::write_address(hal::byte p_address,
+                                 function_ref<hal::timeout_function> p_timeout)
+{
+  // write the address
+  auto acknowledged = write_byte(p_address, p_timeout);
+
+  if (!acknowledged) {
+    hal::safe_throw(hal::no_such_device((p_address >> 1), this));
+  }
+}
+
+void i2c_bit_bang::write(std::span<const hal::byte> p_data_out,
+                         function_ref<hal::timeout_function> p_timeout)
+{
+  bool acknowledged;
+  for (const hal::byte& data : p_data_out) {
+
+    acknowledged = write_byte(data, p_timeout);
+
+    if (!acknowledged) {
+      hal::safe_throw(hal::io_error(this));
+    }
+  }
+}
+
+bool i2c_bit_bang::write_byte(hal::byte p_byte_to_write,
+                              function_ref<hal::timeout_function> p_timeout)
+{
+  using namespace std::chrono_literals;
+  constexpr auto byte_length = 8;
+  constexpr hal::byte bit_select = 0b0000'0001;
+  hal::byte bit_to_write = 0;
+  uint32_t i;
+  for (i = 0; i < byte_length; i++) {
+
+    bit_to_write = static_cast<hal::byte>(
+      (p_byte_to_write >> (byte_length - (i + 1))) & bit_select);
+
+    write_bit(bit_to_write, p_timeout);
+  }
+
+  // look for the ack
+  auto ack_bit = read_bit();
+  // if ack bit is 0, then it was acknowledged (true)
+  return (ack_bit == 0);
```

Comment:
```suggestion
  return ack_bit == 0;
```

Try and avoid using `()` here. It ends up looking like a function call.